core: get infra instead of load the infra in simulation and results

younesschrifi opened the pull request:
Close #4341 
close #3185 
close #4447

leovalais commented on editoast/src/models/timetable.rs:
```diff
@@ -64,6 +74,66 @@ impl Timetable {
+                        .expect("Train should have at least one position")
+                        .time
+                        + train_schedule.departure_time;
+                    let eco = if let Some(eco) = result_train_eco {
+                        Some(eco.0.mechanical_energy_consumed)
+                    } else {
+                        None
+                    };
```
```suggestion
                    let eco = result_train_eco.map(|eco| eco.0.mechanical_energy_consumed);
```

leovalais commented on editoast/src/models/timetable.rs:
```diff
@@ -64,6 +74,66 @@ impl Timetable {
+                    let filtered_stops: Vec<ResultStops> = result_train
+                        .stops
+                        .into_iter()
+                        .filter(|stop| stop.duration > 0 as f64)
```
```suggestion
                        .filter(|stop| stop.duration > 0.)
```

leovalais commented on editoast/src/views/mod.rs:
```diff
@@ -35,8 +35,8 @@ pub fn routes() -> impl HttpServiceFactory {
         light_rolling_stocks::routes(),
-        pathfinding::routes(),
         train_schedule::routes(),
+        pathfinding::routes(),
```
You probably can revert this change 😁

leovalais commented on editoast/src/models/timetable.rs:
```diff
@@ -64,6 +74,66 @@ impl Timetable {
 
+    /// Retrieves timetable with a specific id, its associated train schedules details and
+    /// some informations about the simulation result
+    pub async fn with_train_schedules_with_details(
```
If `Timetable::with_train_schedules` is not used anymore better delete it.

leovalais commented on editoast/src/views/train_schedule/mod.rs:
```diff
@@ -66,17 +64,13 @@ pub enum TrainScheduleError {
 
 /// Return a specific timetable with its associated schedules
-#[get("")]
+#[get("/{id}")]
```
I think it's not necessary to add the `{id}` here since it's induced by the `scope()` above.

leovalais commented on editoast/src/views/train_schedule/simulation_report.rs:
```diff
@@ -95,7 +95,6 @@ pub async fn create_simulation_report(
     } else {
         None
     };
-
```
Either remove this change or also delete the other line breaks in the function.

leovalais commented on editoast/src/models/train_schedule.rs:
```diff
@@ -115,6 +115,22 @@ pub struct TrainScheduleDetails {
+}
+
+#[derive(Serialize, Debug, PartialEq, Clone, Queryable)]
+pub struct TrainScheduleWithDetails {
```
The structs `TrainScheduleDetails` and `TrainScheduleWithDetails` almost have the same name but different contents/usages. Maybe rename one of them and add a small doc comment to explain how/where they're supposed to be used and what they mean.

leovalais reviewed (changes requested):
I have proof read the back only, is that PR still a draft of can I test it now?

multun commented on editoast/openapi.yaml:
```diff
@@ -1441,19 +1441,10 @@ paths:
                   name:
                     type: string
-                  train_schedules:
+                  train_schedules_with_details:
```
```suggestion
                  detailed_train_schedules:
```

leovalais commented:
Just tested GET /timetable/X

leovalais reviewed (approved):
LGTM (editoast)

Khoyo commented on core/src/main/java/fr/sncf/osrd/api/InfraManager.java:
```diff
@@ -160,6 +160,7 @@ private FullInfra downloadInfra(
                     throw new UnexpectedHttpResponse(response);
 
                 // Parse the response
+
```
Spurious newline

flomonster commented on core/src/main/java/fr/sncf/osrd/api/SignalProjectionEndpoint.java:
```diff
@@ -52,10 +52,9 @@ public Response act(Request req) throws Exception {
-            // load infra
-            // TODO : change with get infra when the front is ready
-            var infra = infraManager.load(request.infra, request.expectedVersion, recorder);
+            System.out.println(request.expectedVersion);
```
We should use logger here

flomonster commented on tests/tests/test_timetable.py:
```diff
@@ -81,14 +93,15 @@ def test_get_timetable(
     timetable_id = timetable_id or small_scenario.timetable
 
     response = requests.get(f"{service_url}timetable/{timetable_id}/")
+    print(response.json())
```
```suggestion
```

eckter replied:
Is there a reason why it's resolved? I would have commented the same thing

leovalais replied:
I fixed it in https://github.com/osrd-project/osrd/commit/279df9656859a3fb90e0e07cf12860a0d5a9cf26 but somehow the commit doesn't show here 🤔

leovalais replied:
Well, now the commits has decided to show himself so I'm resolving 😁

flomonster reviewed (approved):
LGTM

nicolaswurtz commented on front/public/locales/fr/simulation.json:
```diff
@@ -2,8 +2,10 @@
   "errorMessages": {
-    "unableToRetrieveTrainSchedule": "Calcul de marche impossible"
+    "unableToRetrieveTrainSchedule": "Calcul de marche impossible",
+    "unableToLoadInfra": "Erreur de chargement de l'infra"
```
If "unable to" it should be translated to « impossible de charger l'infrastructure »
Unless it's an error, so key should be `errorLoadingInfra`
THE CHOICE IS YOURS :D

nicolaswurtz replied:
Infrastructure en toutes lettres ;)

SharglutDev replied:
done :)

eckter commented on tests/tests/test_timetable.py:
```diff
@@ -14,14 +14,27 @@ class _TrainScheduleDetails:
+    timetable_id: int
+    rolling_stock_id: int
+    initial_speed: int
+    labels: Iterable[str]
+    allowances: Iterable[dict]
+    speed_limit_tags: str
+    comfort: str
+    options: dict
+    power_restriction_ranges: Iterable[dict]
+    mechanical_energy_consumed: Iterable[dict]
```
The `Iterables` don't generally support `len(x)`, which we use later on. It should probably be changed to `List`

eckter commented on tests/tests/test_timetable.py:
```diff
@@ -98,16 +111,51 @@ def test_get_timetable(
     expected_schedules = [
-        _TrainScheduleDetails(simulation.ids[0], "West to South East 1", 3600, west_to_south_east_path.id),
-        _TrainScheduleDetails(simulation.ids[1], "West to South East 3", 5100, west_to_south_east_path.id),
+        _TrainScheduleDetails(
+            simulation.ids[0],
+            "West to South East 1",
+            3600,
+            4384.388061579815,
+            west_to_south_east_path.id,
+            small_scenario.timetable,
+            fast_rolling_stock,
+            0.0,
+            ["new train", "west", "south east"],
+            [],
+            None,
+            "STANDARD",
+            None,
+            None,
+            {"base": 6041639237.0, "eco": None},
+            0,
+            45549.5653000392,
+        ),
```
Some types don't match, it causes some warnings in pycharm.

The int/float and None are not that important, but the "expected `Iterable[Dict]`, got `Dict[...]` instead" is more worrying

leovalais replied:
Those should be fixed now.

eckter commented on tests/tests/test_timetable.py:
```diff
@@ -14,14 +14,27 @@ class _TrainScheduleDetails:
     id: int
     name: str
-    train_schedules: Iterable[_TrainScheduleDetails]
+    detailed_train_schedules: Iterable[_TrainScheduleDetails]
```
At this point it's still an iterable of dicts, it's converted to `_TrainScheduleDetails` later on

eckter reviewed (commented):
I only reviewed core and tests, and only tested the tests

eckter reviewed (approved):
LGTM on the core and test changes

Khoyo commented on editoast/src/models/train_schedule.rs:
```diff
@@ -107,14 +107,32 @@ impl From<TrainScheduleChangeset> for TrainSchedule {
+
+/// Returns the timetable with the train's important route information (distance travelled, arrival time, etc.)
+#[derive(Serialize, Debug, PartialEq, Clone, Queryable)]
+pub struct DetailedTrainSchedule {
```
TrainScheduleDetails wasn't a good name, but I think this is worse. It's not detailed, it's a summary (a real TrainSchedule has *more details*)

The timetable is detailed, not the train schedule.

leovalais replied:
Is `TrainScheduleSummary` better?

Khoyo replied:
You got my drift ;)

nicolaswurtz commented on front/src/applications/operationalStudies/components/Scenario/Timetable.tsx:
```diff
@@ -148,25 +159,68 @@ export default function Timetable({
   useEffect(() => {
-    setTrainsDurationsIntervals(findTrainsDurationsIntervals(departureArrivalTimes));
-  }, [departureArrivalTimes]);
+    if (timetable && timetable.train_schedule_summaries) {
+      const scheduledTrains = timetable.train_schedule_summaries;
+
+      const trainsListWithFilterAndDuration = scheduledTrains.map((train: ScheduledTrain) => ({
+        ...train,
+        isFiltered: isTrainFiltered(train, debouncedTerm),
+        duration: durationInSeconds(train.departure_time, train.arrival_time),
+      }));
+      setTrainsList(trainsListWithFilterAndDuration);
+    }
+  }, [trainsDurationsIntervals, debouncedTerm]);
+
+  useEffect(() => {
+    if (timetable && timetable.train_schedule_summaries) {
+      setTrainsDurationsIntervals(findTrainsDurationsIntervals(timetable.train_schedule_summaries));
+    }
+  }, [timetable]);
+
+  useEffect(() => {
+    if (timetableID) {
+      getTimetableWithTrainSchedulesDetails({ id: timetableID })
+        .unwrap()
+        .then((result) => {
+          setTimetable(result);
+        });
+    }
+  }, [timetableID]);
+
+  useEffect(() => {
+    if (timetableID && reloadTimetable) {
+      getTimetableWithTrainSchedulesDetails({ id: timetableID })
+        .unwrap()
+        .then((result) => {
+          setTimetable(result);
+        });
+    }
+  }, [timetableID, reloadTimetable]);
```
You could simplify and avoid 3 useEffect I think, try to fire `setTrainsDurationsIntervals` when updating timetable after fetch, use conditions to test if `reloadTimetable` is needed, etc.

SharglutDev replied:
done :) I move some of the logic from Scenario to Timetable. Cleaner and seems to work properly

nicolaswurtz commented on front/src/applications/operationalStudies/components/SimulationResults/simulationResultsHelpers.jsx:
```diff
@@ -36,8 +38,12 @@ export async function changeTrain(details, id) {
       };
+      dispatch(updateReloadTimetable(true));
       await patch(`${trainscheduleURI}${id}/`, params);
+      dispatch(updateReloadTimetable(false));
```
why put this here ? you should do this IN timetable updating functions

SharglutDev replied:
Hopefully this call can be removed later when we will be using rtk query tags.

SharglutDev replied:
done :) Add the dispatch in duplicateTrain, deleteTrain, SubmitConfAdd and SubmitConfUpdate (transformed it into a component like SubmitAdd to use rtk in it but not with editoast as trainschedules endpoints are broken in it)

SharglutDev replied:
Not sure why I had to do it with SubmitConfUpdate and not SubmitConfAdd (which works the same), but I had to put the dispatch in a Promise.resolve()

nicolaswurtz commented on front/src/applications/operationalStudies/views/Scenario.tsx:
```diff
@@ -97,25 +100,45 @@ export default function Scenario() {
     }
   }, [infra]);
 
+  useEffect(() => {
+    if (infra && infra.state === 'CACHED' && timetableId) {
+      getTimetableWithTrainSchedulesDetails({ id: timetableId })
+        .unwrap()
+        .then((result) => {
+          getTimetable(result);
+        });
+
+      getTimetableConflicts({ id: timetableId })
+        .unwrap()
+        .then((data) => {
+          setConflicts(data as Conflict[]);
+        });
+    }
+  }, [infra, timetableId]);
+
+  useEffect(() => {
+    if (timetableId && reloadTimetable) {
+      getTimetableConflicts({ id: timetableId })
+        .unwrap()
+        .then((data) => {
+          setConflicts(data as Conflict[]);
+        });
+    }
+  }, [reloadTimetable]);
+
+  useEffect(() => {
+    if (infraId) {
+      reloadInfra({ id: infraId as number }).unwrap();
+    }
+  }, [infraId]);
+
```
Same here, it's not possible to make it more concise ? less useEffect ! :D

SharglutDev replied:
Yeah, I'm not a fan either for these 3 but : 
- If i put the 3rd one in the 1st one. It means each time infra will change (and during infra load, polling refetch infra every second so it changes almost every second), it will relaunch the reloadInfra so we will have 2 calls (the reload and the polling) instead of one. The 3rd is actually just here when we enter the page to check if infra is still up to date or if it need to be reloaded.
- If i put the 2nd one in the 1st one, it means each time reloadTimetalbe will change (it always goes from false > true > false when adding/deleting/modifying a train), it will launch multiple api calls for both endpoint in the first useEffect.

If you have a better suggestion/workflow for that, I'd be very happy to try it :)

SharglutDev replied:
Probably we will be able to remove the 2nd one when using rtk query tags

SharglutDev replied:
done :) Moved some of the logic in Timetable

nicolaswurtz commented on front/src/applications/operationalStudies/views/SimulationResults.tsx:
```diff
@@ -102,13 +109,21 @@ export default function SimulationResults({
-      <h1 className="text-center">{t('simulation:noData')}</h1>
-    ) : null;
+  const waitingMessage = () => {
+    if (infraState === 'ERROR' || infraState === 'TRANSIENT_ERROR') {
+      return <h1 className="text-center">{t('simulation:errorMessages.unableToLoadInfra')}</h1>;
+    }
+    if (infraState !== 'CACHED') {
+      return <h1 className="text-center">{t('simulation:infraLoading')}</h1>;
+    }
+    if ((!simulation || simulation.trains.length === 0) && !isUpdating && infraState === 'CACHED') {
+      return <h1 className="text-center">{t('simulation:noData')}</h1>;
+    }
+    return null;
+  };
 
   if (!displaySimulation || isUpdating) {
-    return <div className="pt-5 mt-5">{waitingLoader}</div>;
+    return <div className="pt-5 mt-5">{waitingMessage()}</div>;
```
When loading, put a spinner like « mise à jour »

SharglutDev replied:
done :) Like you suggested, I moved the ScenarioLoader in SimulationResults as well as the text files from scenario.json to simulation.json

nicolaswurtz commented on front/src/applications/stdcm/views/OSRDCStdcmConfig.tsx:
```diff
@@ -36,18 +37,37 @@ export default function OSRDConfig({
   const shouldDisplayStdcmResult = simulation.trains[selectedTrain] !== undefined;
 
+  useEffect(() => {
+    if (infra && infra.state === 'NOT_LOADED') {
+      setIsInfraLoaded(false);
+    }
+
+    if (infra && infra.state === 'CACHED') {
+      setIsInfraLoaded(true);
+    }
```
you should use an enum to list `infra.state` states common for all test about this

SharglutDev replied:
There is already an enum for the state property sent by the back end (can be seen in openapi).
I can add one in the front if you want but maybe for a fix later in another issue ?

nicolaswurtz replied:
You're right, enum lacks from openapi and it's where it should be, so see you in the future !

nicolaswurtz commented on front/src/applications/stdcm/views/OSRDCStdcmConfig.tsx:
```diff
@@ -73,6 +93,7 @@ export default function OSRDConfig({
                 <button
                   className="btn btn-sm  btn-primary "
                   type="button"
+                  disabled={infra?.state !== 'CACHED'}
```
same here

SharglutDev replied:
See comment above

SharglutDev committed:
core: require infra to be load for simulation

SharglutDev committed:
editoast: add information to simulation results

SharglutDev committed:
tests: adapt timetable tests

SharglutDev committed:
front: block stdcm when infra is loading

SharglutDev committed:
front: move infra loader style in scenario

SharglutDev committed:
front: block simulations results

use new editoast timetable endpoint and block simulation results when
infra is loading

SharglutDev committed:
front: trains are now selected by id

clarani commented on front/src/applications/operationalStudies/components/SimulationResults/SimulationResultsMap.tsx:
```diff
@@ -69,12 +71,17 @@ import { MapLayerMouseEvent } from '../../../../types';
   positionValues: PositionValues,
   allowancesSettings?: AllowancesSettings,
-  id?: number,
+  // id?: number,
```
```suggestion
```

SharglutDev replied:
done :)

clarani commented on front/src/applications/operationalStudies/components/SimulationResults/SimulationResultsMap.tsx:
```diff
@@ -69,12 +71,17 @@ import { MapLayerMouseEvent } from '../../../../types';
-    allowancesSettings && id && allowancesSettings[id]?.ecoBlocks ? `eco_${baseKey}` : baseKey
-  ) as keyof PositionValues;
+  const key =
+    // allowancesSettings && id && allowancesSettings[id]?.ecoBlocks ? `eco_${baseKey}` : baseKey
```
```suggestion
```

SharglutDev replied:
done :)

clarani commented on front/src/applications/operationalStudies/components/SimulationResults/SpaceTimeChart/withOSRDData.jsx:
```diff
@@ -103,8 +96,8 @@ const withOSRDData = (Component) =>
 
-    const dispatchUpdateSelectedTrain = (_selectedTrain) => {
-      dispatch(updateSelectedTrain(_selectedTrain));
+    const dispatchUpdateSelectedTrainId = (_selectedTrain) => {
+      dispatch(updateSelectedTrainId(_selectedTrain));
```
```suggestion
    const dispatchUpdateSelectedTrainId = (_selectedTrainId) => {
      dispatch(updateSelectedTrainId(_selectedTrainId));
```

SharglutDev replied:
done :)

clarani commented on front/src/applications/stdcm/components/OldAllowances/EmptyLine.tsx:
```diff
@@ -32,40 +31,37 @@ function getAllowanceValue(values: RangeAllowance | EngineeringAllowance) {
+
+  if (selectedTrain) {
+    if (allowanceType === 'engineering') {
+      // eslint-disable-next-line @typescript-eslint/ban-ts-comment
+      // @ts-ignore
```
```suggestion
```

SharglutDev replied:
I get an error I remove these lines

clarani commented on front/src/applications/stdcm/views/StdcmRequestModal.jsx:
```diff
@@ -79,18 +79,16 @@ export default function StdcmRequestModal(props) {
-            dispatch(updateSimulation(newSimulation));
-            dispatch(updateSelectedTrain(newSimulation.trains.length - 1));
+            dispatch(updateSimulation({ trains }));
+            dispatch(updateSelectedTrainId(trains[0].id));
```
This is for STDCM projection :)
```suggestion
            dispatch(updateSelectedTrainId(fakeNewTrain.id));
```

SharglutDev replied:
done ! Nicely spotted :)

clarani commented on front/src/common/api/osrdEditoastApi.ts:
```diff
@@ -1678,6 +1670,87 @@ export type ProjectPatchRequest = {
+export type TrainScheduleSummary = {
+  id: number;
+  train_name: string;
+  timetable_id?: number;
+  rolling_stock_id?: number;
+  departure_time: number;
+  path_id?: number;
+  initial_speed?: number;
+  labels: string[];
+  allowances?: Allowance[];
+  speed_limit_tags?: string;
+  comfort?: Comfort;
+  options?: TrainScheduleOptions | null;
+  power_restriction_ranges?: PowerRestrictionRange[] | null;
+  arrival_time: number;
+  mechanical_energy_consumed: {
+    base?: number;
+    eco?: number;
+  };
+  stops_count?: number;
+  path_length?: number;
```
Is it normal that all these fields are possibly undefined ?

SharglutDev replied:
I don't think so, but it should be checked with what the back returns ?

clarani commented on front/src/reducers/osrdsimulation/index.ts:
```diff
@@ -155,7 +160,9 @@ export default function reducer(inputState: OsrdSimulationState | undefined, act
         draft.displaySimulation =
           draft.simulation.present?.trains.length > 0 &&
-          draft.simulation.present.trains[state.selectedTrain] !== undefined;
+          // draft.simulation.present.trains[state.selectedTrain] !== undefined;
```
```suggestion
```

SharglutDev replied:
done :)

nicolaswurtz commented on front/src/applications/operationalStudies/components/SimulationResults/SpaceTimeChart/createTrain.ts:
```diff
@@ -75,41 +75,37 @@ export default function createTrain(
+    signalAspects: formatSignalAspects(train.base.signal_aspects),
+    speed: formatStepsWithTime(train.base.speeds),
+    // trainNumber is not used anymore now that we use the train id, but it is still required for TS
+    trainNumber: 0,
```
Find where it's needed and remove it please !
